internal/provider/resource_storagezone.go

```diff
@@ -0,0 +1,397 @@
+}
+
+func immutableStringPropertyError(key string, old string, new string) error {
+	message := "'%s' is immutable and cannot be changed from '%s' to '%s'. " +
+		"If you must change the '%s' of our region you must first delete your resource and then redefine it. " +
+		"WARNING: deleting a 'bunny_storagezone' will also delete all the data it contains!"
+	return fmt.Errorf(message, key, old, new, key)
+}
+
+func immutableReplicationRegionError(key string, removed []interface{}) error {
+	message := "'%s' can be added to but not be removed once the zone has been created. " +
+		"This error occurred when attempting to remove values %+q from '%s'. " +
+		"To remove an existing '%s' the 'bunny_storagezone' must be deleted and recreated. " +
+		"WARNING: deleting a 'bunny_storagezone' will also delete all the data it contains!"
+	return fmt.Errorf(
+		message,
+		key,
+		removed,
+		key,
+		key,
+	)
+}
+
+func resourceStorageZoneCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	clt := meta.(*bunny.Client)
+
+	originURL := getStrPtr(d, keyOriginURL)
+	if !d.HasChange(keyOriginURL) {
+		originURL = nil
+	}
+
+	sz, err := clt.StorageZone.Add(ctx, &bunny.StorageZoneAddOptions{
+		Name:               getStrPtr(d, keyName),
+		OriginURL:          originURL,
+		Region:             getStrPtr(d, keyRegion),
+		ReplicationRegions: getStrSetAsSlice(d, keyReplicationRegions),
+	})
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("creating storage zone failed: %w", err))
+	}
+
+	d.SetId(strconv.FormatInt(*sz.ID, 10))
+
+	// StorageZone.Add() only supports to set a subset of a Storage Zone object,
+	// call Update to set the remaining ones.
+	if diags := resourceStorageZoneUpdate(ctx, d, meta); diags.HasError() {
+		// if updating fails the sz was still created, initialize with the SZ
+		// returned from the Add operation
+		diags = append(diags, diag.Diagnostic{
+			Severity: diag.Error,
+			Summary:  "setting storage zone attributes via update failed",
+		})
+
+		if err := storageZoneToResource(sz, d); err != nil {
+			diags = append(diags, diag.Diagnostic{
+				Severity: diag.Error,
+				Summary:  "converting api-type to resource data failed: " + err.Error(),
+			})
+		}
+
+		return diags
+	}
+
+	return nil
+}
+
+func resourceStorageZoneUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	clt := meta.(*bunny.Client)
+
+	storageZone, err := storageZoneFromResource(d)
+	if err != nil {
+		return diagsErrFromErr("converting resource to API type failed", err)
+	}
+
+	id, err := getIDAsInt64(d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	updateErr := clt.StorageZone.Update(ctx, id, storageZone)
+	if updateErr != nil {
+		// if our update failed then revert our values to their original
+		// state so that we can run an apply again.
+		revertErr := revertUpdateValues(d)
+
+		if revertErr != nil {
+			return diagsErrFromErr("updating storage zone via API failed", revertErr)
+		}
```

Comment:
This is not needed, the d.Get calls returns the proposed data.
The current resource state is never changed.